src/amoc_metrics.erl

```diff
@@ -1,66 +1,96 @@
+    #{reporter := Reporter, interval := Interval} = Opts,
+    Datapoints = maps:get(report, Opts, [one, count]),
+    create_metric_and_maybe_subscribe(ExName, spiral, Reporter, Interval, Datapoints).
 
 -spec new_counter(name()) -> ok.
 new_counter(Name) when is_atom(Name) ->
-    new_counter(Name, []).
+    new_counter(Name, default_reporting_opts()).
 
--spec new_counter(name(), [proplists:property()]) -> ok.
+-spec new_counter(name(), reporting_opts()) -> ok.
 new_counter(Name, Opts) when is_atom(Name) ->
     ExName = [amoc, counters, Name],
-    {Reporter, Interval} = reporter_and_interval(Opts),
-    ReportProps = proplists:get_value(report, Opts, [value]),
-    ok = exometer:new(ExName, counter),
-    ok = exometer_report:subscribe(Reporter, ExName, ReportProps, Interval).
+    #{reporter := Reporter, interval := Interval} = Opts,
+    Datapoints = maps:get(report, Opts, [value]),
+    create_metric_and_maybe_subscribe(ExName, counter, Reporter, Interval, Datapoints).
 
 -spec update_hist(name(), integer()) -> ok.
-update_hist(Name, Value) when is_atom(Name) ->
-    ExName = [amoc, histograms, Name],
+update_hist(Name, Value) ->
+    ExName = make_name(histogram, Name),
     exometer:update(ExName, Value).
 
 -spec update_spiral(name(), integer()) -> ok.
-update_spiral(Name, Value) when is_atom(Name) ->
-    ExName = [amoc, spirals, Name],
+update_spiral(Name, Value) ->
+    ExName = make_name(spiral, Name),
     exometer:update(ExName, Value).
 
 -spec update_counter(name(), integer()) -> ok.
-update_counter(Name, Value) when is_atom(Name) ->
-    ExName = [amoc, counters, Name],
+update_counter(Name, Value) ->
+    ExName = make_name(counter, Name),
     exometer:update(ExName, Value).
 
--spec reporter_and_interval([proplists:property()]) ->
-    {module(), non_neg_integer()}.
-reporter_and_interval(Opts) ->
-    Reporter = proplists:get_value(reporter, Opts, exometer_report_graphite),
-    Interval = proplists:get_value(interval, Opts, 5000),
-    {Reporter, Interval}.
+-spec messages_spiral_name() -> name().
+messages_spiral_name() ->
+    [counters, messages_sent].
+
+-spec message_ttd_histogram_name() -> name().
+message_ttd_histogram_name() ->
+    [times, message_ttd].
+
+-spec default_reporting_opts() -> reporting_opts().
+default_reporting_opts() ->
+    #{reporter => exometer_report_graphite,
+      interval => timer:seconds(10)}.
+
+make_name(Type, Name) when is_atom(Name) ->
+    [amoc, name_by_type(Type), Name];
+make_name(_, Name) when is_list(Name) ->
+    [amoc | Name].
+
+name_by_type(histogram) -> histograms;
+name_by_type(spiral) -> spirals;
+name_by_type(counter) -> counters.
+
+
+create_metric_and_maybe_subscribe(ExName, Type, Reporter, Interval, Datapoints) ->
+    ok = exometer:re_register(ExName, Type, []),
+    case lists:keyfind(Reporter, 1, exometer_report:list_reporters()) of
+        {Reporter, _} ->
+            exometer_report:unsubscribe(Reporter, ExName, Datapoints),
```

Comment:
I am not sure if we need to unsubscribe here. Will this fail when already subscribed?

Reply:
When already subscribed and we subscribe again we may end up with more than one subscription sending the same metric on different times.